### ISSUE TITLE
ManifoldCallback example with lower tolerances randomly converges or not

### ISSUE THREAD

**Issue by abavoil** [SciML/DiffEqCallbacks.jl#244]
**Describe the example**

I tried reproducing the code found [here](https://docs.sciml.ai/DiffEqCallbacks/stable/projection/) with different tolerances and came up with the code below. **Everytime I execute this script I get a different result, solves with comeback will randomly converge or not**

**Minimal Reproducible Example 👇**

```julia
using Pkg
Pkg.activate(; temp=true)
Pkg.add(name="NonlinearSolve", version="3.15.1")
Pkg.add(name="DiffEqCallbacks", version="4.0.0")
Pkg.add(name="OrdinaryDiffEq", version="6.90.1")
Pkg.add("ADTypes")

using OrdinaryDiffEq
using DiffEqCallbacks
using LinearAlgebra: norm
using NonlinearSolve
using ADTypes: AutoForwardDiff

function f!(du, u, p, t)
    du .= [-u[2], u[1]]
end

function g!(resid, u, p, t)
    resid[1] = norm(u) - 1
end

u0 = [1., 0.]
tf = 1e3
for tol in (1e-3, 1e-6, 1e-9, 1e-12, 1e-15)
    prob = ODEProblem(f!, u0, tf)
    cb = ManifoldProjection(g!, save=false, autodiff=AutoForwardDiff(), resid_prototype=zeros(eltype(u0), 1))
    sol = solve(prob, Tsit5(), reltol=tol, abstol=tol)
    sol_cb = solve(prob, Tsit5(), reltol=tol, abstol=tol, callback=cb)
    @show tol
    @show sol.retcode
    @show sol_cb.retcode
    println("resid    = ", g!(zeros(1), sol.u[end], nothing, sol.t[end]))
    println("resid_cb = ", g!(zeros(1), sol_cb.u[end], nothing, sol_cb.t[end]))
    println()
end
```

**Error & Stacktrace ⚠️**

No error, just a stochastic behavior: I run the script a few times and get different results.

**Not Working Environment (please complete the following information):**

  - Output of `using Pkg; Pkg.status()`

```julia
  [47edcb42] ADTypes v1.11.0
⌃ [459566f4] DiffEqCallbacks v4.0.0
⌃ [8913a72c] NonlinearSolve v3.15.1
  [1dea7af3] OrdinaryDiffEq v6.90.1
```

  - Output of `using Pkg; Pkg.status(; mode = PKGMODE_MANIFEST)`
<details>
  <summary>Manifest</summary>

```julia  [47edcb42] ADTypes v1.11.0
  [7d9f7c33] Accessors v0.1.39
  [79e6a3ab] Adapt v4.1.1
  [ec485272] ArnoldiMethod v0.4.0
  [4fba245c] ArrayInterface v7.17.1
  [4c555306] ArrayLayouts v1.11.0
  [62783981] BitTwiddlingConvenienceFunctions v0.1.6
  [2a0fbf3d] CPUSummary v0.2.6
  [d360d2e6] ChainRulesCore v1.25.0
  [fb6a15b2] CloseOpenIntervals v0.1.13
  [38540f10] CommonSolve v0.2.4
  [bbf7d656] CommonSubexpressions v0.3.1
  [f70d9fcc] CommonWorldInvalidations v1.0.0
  [34da2185] Compat v4.16.0
  [a33af91c] CompositionsBase v0.1.2
  [2569d6c7] ConcreteStructs v0.2.3
  [187b0558] ConstructionBase v1.5.8
  [adafc99b] CpuId v0.3.1
  [9a962f9c] DataAPI v1.16.0
  [864edb3b] DataStructures v0.18.20
  [e2d170a0] DataValueInterfaces v1.0.0
  [2b5f629d] DiffEqBase v6.160.0
⌃ [459566f4] DiffEqCallbacks v4.0.0
  [163ba53b] DiffResults v1.1.0
  [b552c78f] DiffRules v1.15.1
  [a0c0ee7d] DifferentiationInterface v0.6.27
  [ffbed154] DocStringExtensions v0.9.3
  [4e289a0a] EnumX v1.0.4
  [f151be2c] EnzymeCore v0.8.8
  [d4d017d3] ExponentialUtilities v1.27.0
  [e2ba6199] ExprTools v0.1.10
⌅ [6b7a57c9] Expronicon v0.8.5
  [7034ab61] FastBroadcast v0.3.5
  [9aa1b823] FastClosures v0.3.2
  [29a986be] FastLapackInterface v2.0.4
  [a4df4552] FastPower v1.1.1
  [1a297f60] FillArrays v1.13.0
  [6a86dc24] FiniteDiff v2.26.2
  [f6369f11] ForwardDiff v0.10.38
  [069b7b12] FunctionWrappers v1.1.3
  [77dc65aa] FunctionWrappersWrappers v0.1.3
⌅ [d9f16b24] Functors v0.4.12
  [46192b85] GPUArraysCore v0.2.0
  [c145ed77] GenericSchur v0.5.4
  [86223c79] Graphs v1.12.0
  [3e5b6fbb] HostCPUFeatures v0.1.17
  [615f187c] IfElse v0.1.1
  [d25df0c9] Inflate v0.1.5
  [3587e190] InverseFunctions v0.1.17
  [92d709cd] IrrationalConstants v0.2.2
  [82899510] IteratorInterfaceExtensions v1.0.0
  [692b3bcd] JLLWrappers v1.6.1
  [ef3ab10e] KLU v0.6.0
  [ba0b0d4f] Krylov v0.9.8
  [10f19ff3] LayoutPointers v0.1.17
  [5078a376] LazyArrays v2.3.1
  [87fe0de2] LineSearch v0.1.4
  [d3d80556] LineSearches v7.3.0
  [7ed4a6bd] LinearSolve v2.38.0
  [2ab3a3ac] LogExpFunctions v0.3.28
  [bdcacae8] LoopVectorization v0.12.171
  [d8e11817] MLStyle v0.4.17
  [1914dd2f] MacroTools v0.5.13
  [d125e4d3] ManualMemory v0.1.8
  [bb5d69b7] MaybeInplace v0.1.4
  [46d2c3a1] MuladdMacro v0.2.4
  [d41bc354] NLSolversBase v7.8.3
  [77ba4419] NaNMath v1.0.2
⌃ [8913a72c] NonlinearSolve v3.15.1
  [6fe1bfb0] OffsetArrays v1.14.2
  [bac558e1] OrderedCollections v1.7.0
  [1dea7af3] OrdinaryDiffEq v6.90.1
  [89bda076] OrdinaryDiffEqAdamsBashforthMoulton v1.1.0
  [6ad6398a] OrdinaryDiffEqBDF v1.1.2
  [bbf590c4] OrdinaryDiffEqCore v1.13.0
  [50262376] OrdinaryDiffEqDefault v1.1.0
  [4302a76b] OrdinaryDiffEqDifferentiation v1.2.0
  [9286f039] OrdinaryDiffEqExplicitRK v1.1.0
  [e0540318] OrdinaryDiffEqExponentialRK v1.1.0
  [becaefa8] OrdinaryDiffEqExtrapolation v1.2.1
  [5960d6e9] OrdinaryDiffEqFIRK v1.5.0
  [101fe9f7] OrdinaryDiffEqFeagin v1.1.0
  [d3585ca7] OrdinaryDiffEqFunctionMap v1.1.1
  [d28bc4f8] OrdinaryDiffEqHighOrderRK v1.1.0
  [9f002381] OrdinaryDiffEqIMEXMultistep v1.1.0
  [521117fe] OrdinaryDiffEqLinear v1.1.0
  [1344f307] OrdinaryDiffEqLowOrderRK v1.2.0
  [b0944070] OrdinaryDiffEqLowStorageRK v1.2.1
  [127b3ac7] OrdinaryDiffEqNonlinearSolve v1.3.0
  [c9986a66] OrdinaryDiffEqNordsieck v1.1.0
  [5dd0a6cf] OrdinaryDiffEqPDIRK v1.1.0
  [5b33eab2] OrdinaryDiffEqPRK v1.1.0
  [04162be5] OrdinaryDiffEqQPRK v1.1.0
  [af6ede74] OrdinaryDiffEqRKN v1.1.0
  [43230ef6] OrdinaryDiffEqRosenbrock v1.3.1
  [2d112036] OrdinaryDiffEqSDIRK v1.1.0
  [669c94d9] OrdinaryDiffEqSSPRK v1.2.0
  [e3e12d00] OrdinaryDiffEqStabilizedIRK v1.1.0
  [358294b1] OrdinaryDiffEqStabilizedRK v1.1.0
  [fa646aed] OrdinaryDiffEqSymplecticRK v1.1.0
  [b1df2697] OrdinaryDiffEqTsit5 v1.1.0
  [79d7bb75] OrdinaryDiffEqVerner v1.1.1
  [65ce6f38] PackageExtensionCompat v1.0.2
  [d96e819e] Parameters v0.12.3
  [f517fe37] Polyester v0.7.16
  [1d0040c9] PolyesterWeave v0.2.2
  [d236fae5] PreallocationTools v0.4.24
  [aea7be01] PrecompileTools v1.2.1
  [21216c6a] Preferences v1.4.3
  [3cdcf5f2] RecipesBase v1.3.4
  [731186ca] RecursiveArrayTools v3.27.4
  [f2c3362d] RecursiveFactorization v0.2.23
  [189a3867] Reexport v1.2.2
  [ae029012] Requires v1.3.0
  [7e49a35a] RuntimeGeneratedFunctions v0.5.13
  [94e857df] SIMDTypes v0.1.0
  [476501e8] SLEEFPirates v0.6.43
  [0bca4576] SciMLBase v2.67.0
  [19f34311] SciMLJacobianOperators v0.1.1
  [c0aeaf25] SciMLOperators v0.3.12
  [53ae85a6] SciMLStructures v1.6.1
  [efcf1570] Setfield v1.1.1
⌅ [727e6d20] SimpleNonlinearSolve v1.12.3
  [699a6c99] SimpleTraits v0.9.4
  [ce78b400] SimpleUnPack v1.1.0
  [9f842d2f] SparseConnectivityTracer v0.6.9
  [47a9eef4] SparseDiffTools v2.23.0
  [0a514795] SparseMatrixColorings v0.4.10
  [e56a9233] Sparspak v0.3.9
  [276daf66] SpecialFunctions v2.5.0
  [aedffcd0] Static v1.1.1
  [0d7ed370] StaticArrayInterface v1.8.0
  [90137ffa] StaticArrays v1.9.8
  [1e83bf80] StaticArraysCore v1.4.3
  [10745b16] Statistics v1.11.1
  [7792a7ef] StrideArraysCore v0.5.7
  [2efcf032] SymbolicIndexingInterface v0.3.36
  [3783bdb8] TableTraits v1.0.1
  [bd369af6] Tables v1.12.0
  [8290d209] ThreadingUtilities v0.5.2
  [a759f4b9] TimerOutputs v0.5.26
  [d5829a12] TriangularSolve v0.2.1
  [781d530d] TruncatedStacktraces v1.4.0
  [3a884ed6] UnPack v1.0.2
  [3d5dd08c] VectorizationBase v0.21.71
  [19fa3120] VertexSafeGraphs v0.2.0
  [1d5cc7b8] IntelOpenMP_jll v2024.2.1+0
  [856f044c] MKL_jll v2024.2.0+0
  [efe28fd5] OpenSpecFun_jll v0.5.5+0
  [1317d2d5] oneTBB_jll v2021.12.0+0
  [0dad84c5] ArgTools v1.1.2
  [56f22d72] Artifacts v1.11.0
  [2a0f44e3] Base64 v1.11.0
  [ade2ca70] Dates v1.11.0
  [8ba89e20] Distributed v1.11.0
  [f43a241f] Downloads v1.6.0
  [7b1f6079] FileWatching v1.11.0
  [9fa8497b] Future v1.11.0
  [b77e0a4c] InteractiveUtils v1.11.0
  [4af54fe1] LazyArtifacts v1.11.0
  [b27032c2] LibCURL v0.6.4
  [76f85450] LibGit2 v1.11.0
  [8f399da3] Libdl v1.11.0
  [37e2e46d] LinearAlgebra v1.11.0
  [56ddb016] Logging v1.11.0
  [d6f4376e] Markdown v1.11.0
  [a63ad114] Mmap v1.11.0
  [ca575930] NetworkOptions v1.2.0
  [44cfe95a] Pkg v1.11.0
  [de0858da] Printf v1.11.0
  [9a3f8284] Random v1.11.0
  [ea8e919c] SHA v0.7.0
  [9e88b42a] Serialization v1.11.0
  [1a1011a3] SharedArrays v1.11.0
  [6462fe0b] Sockets v1.11.0
  [2f01184e] SparseArrays v1.11.0
  [fa267f1f] TOML v1.0.3
  [a4e569a6] Tar v1.10.0
  [8dfed614] Test v1.11.0
  [cf7118a7] UUIDs v1.11.0
  [4ec0a83e] Unicode v1.11.0
  [e66e0078] CompilerSupportLibraries_jll v1.1.1+0
  [deac9b47] LibCURL_jll v8.6.0+0
  [e37daf67] LibGit2_jll v1.7.2+0
  [29816b5a] LibSSH2_jll v1.11.0+1
  [c8ffd9c3] MbedTLS_jll v2.28.6+0
  [14a3606d] MozillaCACerts_jll v2023.12.12
  [4536629a] OpenBLAS_jll v0.3.27+1
  [05823500] OpenLibm_jll v0.8.1+2
  [bea87d4a] SuiteSparse_jll v7.7.0+0
  [83775a58] Zlib_jll v1.2.13+1
  [8e850b90] libblastrampoline_jll v5.11.0+0
  [8e850ede] nghttp2_jll v1.59.0+0
  [3f19e933] p7zip_jll v17.4.0+2
```
</details>
  - Output of `versioninfo()`

```julia
Julia Version 1.11.2
Commit 5e9a32e7af2 (2024-12-01 20:02 UTC)
Build Info:
  Official https://julialang.org/ release
Platform Info:
  OS: macOS (arm64-apple-darwin24.0.0)
  CPU: 8 × Apple M2
  WORD_SIZE: 64
  LLVM: libLLVM-16.0.6 (ORCJIT, apple-m2)
Threads: 1 default, 0 interactive, 1 GC (on 4 virtual cores)
```

**Working Environment (please complete the following information):**

  - Output of `using Pkg; Pkg.status()`

```julia
⌃ [459566f4] DiffEqCallbacks v3.9.1
⌅ [8913a72c] NonlinearSolve v3.15.1
  [1dea7af3] OrdinaryDiffEq v6.90.1
```

  - Output of `using Pkg; Pkg.status(; mode = PKGMODE_MANIFEST)`

<details>
  <summary>Manifest</summary>
```julia
  [47edcb42] ADTypes v1.11.0
  [7d9f7c33] Accessors v0.1.39
  [79e6a3ab] Adapt v4.1.1
  [ec485272] ArnoldiMethod v0.4.0
  [4fba245c] ArrayInterface v7.17.1
  [4c555306] ArrayLayouts v1.11.0
  [62783981] BitTwiddlingConvenienceFunctions v0.1.6
  [2a0fbf3d] CPUSummary v0.2.6
  [d360d2e6] ChainRulesCore v1.25.0
  [fb6a15b2] CloseOpenIntervals v0.1.13
  [38540f10] CommonSolve v0.2.4
  [bbf7d656] CommonSubexpressions v0.3.1
  [f70d9fcc] CommonWorldInvalidations v1.0.0
  [34da2185] Compat v4.16.0
  [a33af91c] CompositionsBase v0.1.2
  [2569d6c7] ConcreteStructs v0.2.3
  [187b0558] ConstructionBase v1.5.8
  [adafc99b] CpuId v0.3.1
  [9a962f9c] DataAPI v1.16.0
  [864edb3b] DataStructures v0.18.20
  [e2d170a0] DataValueInterfaces v1.0.0
  [2b5f629d] DiffEqBase v6.160.0
⌃ [459566f4] DiffEqCallbacks v3.9.1
  [163ba53b] DiffResults v1.1.0
  [b552c78f] DiffRules v1.15.1
  [a0c0ee7d] DifferentiationInterface v0.6.27
  [ffbed154] DocStringExtensions v0.9.3
  [4e289a0a] EnumX v1.0.4
  [f151be2c] EnzymeCore v0.8.8
  [d4d017d3] ExponentialUtilities v1.27.0
  [e2ba6199] ExprTools v0.1.10
⌅ [6b7a57c9] Expronicon v0.8.5
  [7034ab61] FastBroadcast v0.3.5
  [9aa1b823] FastClosures v0.3.2
  [29a986be] FastLapackInterface v2.0.4
  [a4df4552] FastPower v1.1.1
  [1a297f60] FillArrays v1.13.0
  [6a86dc24] FiniteDiff v2.26.2
  [f6369f11] ForwardDiff v0.10.38
  [069b7b12] FunctionWrappers v1.1.3
  [77dc65aa] FunctionWrappersWrappers v0.1.3
⌅ [d9f16b24] Functors v0.4.12
  [46192b85] GPUArraysCore v0.2.0
  [c145ed77] GenericSchur v0.5.4
  [86223c79] Graphs v1.12.0
  [3e5b6fbb] HostCPUFeatures v0.1.17
  [615f187c] IfElse v0.1.1
  [d25df0c9] Inflate v0.1.5
  [3587e190] InverseFunctions v0.1.17
  [92d709cd] IrrationalConstants v0.2.2
  [82899510] IteratorInterfaceExtensions v1.0.0
  [692b3bcd] JLLWrappers v1.6.1
  [ef3ab10e] KLU v0.6.0
  [ba0b0d4f] Krylov v0.9.8
  [10f19ff3] LayoutPointers v0.1.17
  [5078a376] LazyArrays v2.3.1
  [87fe0de2] LineSearch v0.1.4
  [d3d80556] LineSearches v7.3.0
  [7ed4a6bd] LinearSolve v2.38.0
  [2ab3a3ac] LogExpFunctions v0.3.28
  [bdcacae8] LoopVectorization v0.12.171
  [d8e11817] MLStyle v0.4.17
  [1914dd2f] MacroTools v0.5.13
  [d125e4d3] ManualMemory v0.1.8
  [bb5d69b7] MaybeInplace v0.1.4
  [46d2c3a1] MuladdMacro v0.2.4
  [d41bc354] NLSolversBase v7.8.3
  [77ba4419] NaNMath v1.0.2
⌅ [8913a72c] NonlinearSolve v3.15.1
  [6fe1bfb0] OffsetArrays v1.14.2
  [bac558e1] OrderedCollections v1.7.0
  [1dea7af3] OrdinaryDiffEq v6.90.1
  [89bda076] OrdinaryDiffEqAdamsBashforthMoulton v1.1.0
  [6ad6398a] OrdinaryDiffEqBDF v1.1.2
  [bbf590c4] OrdinaryDiffEqCore v1.13.0
  [50262376] OrdinaryDiffEqDefault v1.1.0
  [4302a76b] OrdinaryDiffEqDifferentiation v1.2.0
  [9286f039] OrdinaryDiffEqExplicitRK v1.1.0
  [e0540318] OrdinaryDiffEqExponentialRK v1.1.0
  [becaefa8] OrdinaryDiffEqExtrapolation v1.2.1
  [5960d6e9] OrdinaryDiffEqFIRK v1.5.0
  [101fe9f7] OrdinaryDiffEqFeagin v1.1.0
  [d3585ca7] OrdinaryDiffEqFunctionMap v1.1.1
  [d28bc4f8] OrdinaryDiffEqHighOrderRK v1.1.0
  [9f002381] OrdinaryDiffEqIMEXMultistep v1.1.0
  [521117fe] OrdinaryDiffEqLinear v1.1.0
  [1344f307] OrdinaryDiffEqLowOrderRK v1.2.0
  [b0944070] OrdinaryDiffEqLowStorageRK v1.2.1
  [127b3ac7] OrdinaryDiffEqNonlinearSolve v1.3.0
  [c9986a66] OrdinaryDiffEqNordsieck v1.1.0
  [5dd0a6cf] OrdinaryDiffEqPDIRK v1.1.0
  [5b33eab2] OrdinaryDiffEqPRK v1.1.0
  [04162be5] OrdinaryDiffEqQPRK v1.1.0
  [af6ede74] OrdinaryDiffEqRKN v1.1.0
  [43230ef6] OrdinaryDiffEqRosenbrock v1.3.1
  [2d112036] OrdinaryDiffEqSDIRK v1.1.0
  [669c94d9] OrdinaryDiffEqSSPRK v1.2.0
  [e3e12d00] OrdinaryDiffEqStabilizedIRK v1.1.0
  [358294b1] OrdinaryDiffEqStabilizedRK v1.1.0
  [fa646aed] OrdinaryDiffEqSymplecticRK v1.1.0
  [b1df2697] OrdinaryDiffEqTsit5 v1.1.0
  [79d7bb75] OrdinaryDiffEqVerner v1.1.1
  [65ce6f38] PackageExtensionCompat v1.0.2
  [d96e819e] Parameters v0.12.3
  [f517fe37] Polyester v0.7.16
  [1d0040c9] PolyesterWeave v0.2.2
  [d236fae5] PreallocationTools v0.4.24
  [aea7be01] PrecompileTools v1.2.1
  [21216c6a] Preferences v1.4.3
  [3cdcf5f2] RecipesBase v1.3.4
  [731186ca] RecursiveArrayTools v3.27.4
  [f2c3362d] RecursiveFactorization v0.2.23
  [189a3867] Reexport v1.2.2
  [ae029012] Requires v1.3.0
  [7e49a35a] RuntimeGeneratedFunctions v0.5.13
  [94e857df] SIMDTypes v0.1.0
  [476501e8] SLEEFPirates v0.6.43
  [0bca4576] SciMLBase v2.67.0
  [19f34311] SciMLJacobianOperators v0.1.1
  [c0aeaf25] SciMLOperators v0.3.12
  [53ae85a6] SciMLStructures v1.6.1
  [efcf1570] Setfield v1.1.1
⌅ [727e6d20] SimpleNonlinearSolve v1.12.3
  [699a6c99] SimpleTraits v0.9.4
  [ce78b400] SimpleUnPack v1.1.0
  [9f842d2f] SparseConnectivityTracer v0.6.9
  [47a9eef4] SparseDiffTools v2.23.0
  [0a514795] SparseMatrixColorings v0.4.10
  [e56a9233] Sparspak v0.3.9
  [276daf66] SpecialFunctions v2.5.0
  [aedffcd0] Static v1.1.1
  [0d7ed370] StaticArrayInterface v1.8.0
  [90137ffa] StaticArrays v1.9.8
  [1e83bf80] StaticArraysCore v1.4.3
  [10745b16] Statistics v1.11.1
  [7792a7ef] StrideArraysCore v0.5.7
  [2efcf032] SymbolicIndexingInterface v0.3.36
  [3783bdb8] TableTraits v1.0.1
  [bd369af6] Tables v1.12.0
  [8290d209] ThreadingUtilities v0.5.2
  [a759f4b9] TimerOutputs v0.5.26
  [d5829a12] TriangularSolve v0.2.1
  [781d530d] TruncatedStacktraces v1.4.0
  [3a884ed6] UnPack v1.0.2
  [3d5dd08c] VectorizationBase v0.21.71
  [19fa3120] VertexSafeGraphs v0.2.0
  [1d5cc7b8] IntelOpenMP_jll v2024.2.1+0
  [856f044c] MKL_jll v2024.2.0+0
  [efe28fd5] OpenSpecFun_jll v0.5.5+0
  [1317d2d5] oneTBB_jll v2021.12.0+0
  [0dad84c5] ArgTools v1.1.2
  [56f22d72] Artifacts v1.11.0
  [2a0f44e3] Base64 v1.11.0
  [ade2ca70] Dates v1.11.0
  [8ba89e20] Distributed v1.11.0
  [f43a241f] Downloads v1.6.0
  [7b1f6079] FileWatching v1.11.0
  [9fa8497b] Future v1.11.0
  [b77e0a4c] InteractiveUtils v1.11.0
  [4af54fe1] LazyArtifacts v1.11.0
  [b27032c2] LibCURL v0.6.4
  [76f85450] LibGit2 v1.11.0
  [8f399da3] Libdl v1.11.0
  [37e2e46d] LinearAlgebra v1.11.0
  [56ddb016] Logging v1.11.0
  [d6f4376e] Markdown v1.11.0
  [a63ad114] Mmap v1.11.0
  [ca575930] NetworkOptions v1.2.0
  [44cfe95a] Pkg v1.11.0
  [de0858da] Printf v1.11.0
  [9a3f8284] Random v1.11.0
  [ea8e919c] SHA v0.7.0
  [9e88b42a] Serialization v1.11.0
  [1a1011a3] SharedArrays v1.11.0
  [6462fe0b] Sockets v1.11.0
  [2f01184e] SparseArrays v1.11.0
  [fa267f1f] TOML v1.0.3
  [a4e569a6] Tar v1.10.0
  [8dfed614] Test v1.11.0
  [cf7118a7] UUIDs v1.11.0
  [4ec0a83e] Unicode v1.11.0
  [e66e0078] CompilerSupportLibraries_jll v1.1.1+0
  [deac9b47] LibCURL_jll v8.6.0+0
  [e37daf67] LibGit2_jll v1.7.2+0
  [29816b5a] LibSSH2_jll v1.11.0+1
  [c8ffd9c3] MbedTLS_jll v2.28.6+0
  [14a3606d] MozillaCACerts_jll v2023.12.12
  [4536629a] OpenBLAS_jll v0.3.27+1
  [05823500] OpenLibm_jll v0.8.1+2
  [bea87d4a] SuiteSparse_jll v7.7.0+0
  [83775a58] Zlib_jll v1.2.13+1
  [8e850b90] libblastrampoline_jll v5.11.0+0
  [8e850ede] nghttp2_jll v1.59.0+0
  [3f19e933] p7zip_jll v17.4.0+2
```
</details>

  - Output of `versioninfo()`

```julia
Julia Version 1.11.2
Commit 5e9a32e7af2 (2024-12-01 20:02 UTC)
Build Info:
  Official https://julialang.org/ release
Platform Info:
  OS: macOS (arm64-apple-darwin24.0.0)
  CPU: 8 × Apple M2
  WORD_SIZE: 64
  LLVM: libLLVM-16.0.6 (ORCJIT, apple-m2)
Threads: 1 default, 0 interactive, 1 GC (on 4 virtual cores)
```

**Additional context**

Here is a working example with v3 where the solves with callbacks all converge, all the time :

```julia
using Pkg

Pkg.activate("tmp/cb3")
Pkg.add(name="NonlinearSolve", version="3.15.1")
Pkg.add(name="DiffEqCallbacks", version="3")
Pkg.add(name="OrdinaryDiffEq", version="6.90.1")

using OrdinaryDiffEq
using DiffEqCallbacks
using LinearAlgebra: norm
using Logging

function f!(du, u, p, t)
    du .= [-u[2], u[1]]
end

function g!(resid, u, p, t)
    resid[2] = 0
    resid[1] = norm(u) - 1
end

with_logger(ConsoleLogger(stderr, 1001)) do
    u0 = [1., 0.]
    tf = 1e3
    for tol in (1e-3, 1e-6, 1e-9, 1e-12, 1e-15)
        prob = ODEProblem(f!, u0, tf)
        cb = ManifoldProjection(g!, save=false)
        sol = solve(prob, Tsit5(), reltol=tol, abstol=tol)
        sol_cb = solve(prob, Tsit5(), reltol=tol, abstol=tol, callback=cb)
        @show tol
        @show sol_cb.retcode
        @show sol_cb.retcode
        println("resid    = ", g!(zeros(2), sol.u[end], nothing, sol.t[end]))
        println("resid_cb = ", g!(zeros(2), sol_cb.u[end], nothing, sol_cb.t[end]))
        println()
    end
end
```

**Comment by abavoil** [comment]
The bug appeared in a script that uses v3.9, but I can't find an MRE.

Basically I have : 
```julia
function g!(resid, y, p)
    resid[1:6] .= something()
    # resid 7:11 are not set to zero
end

y0 = Float64[1:11...]
cb = ManifoldProjection(g!);
prob = ODEProblem(f!!, y0, tspan, p)
solve(prob,, callback=cb)
```
Here, the solve succeeds a third/fourth of the time. But if I modify `g!` as : 
```julia
function g!(resid, y, p)
    resid .= zero(eltype(resid))
    resid[1:6] .= something()
end
```
then the issue is gone.

In the MRE of my initial post, I observed a very similar behavior (the same?) but I couldn't find a way to fix that oneone.

**Comment by ChrisRackauckas** [comment]
If you don't set the residual values to zero, then it's not guaranteed that the residual terms are zero. I don't see the issue there? 

Though in this case, you want to set the number of residual terms.

> residual_prototype: The size of the manifold residual. If it is not specified, we assume it to be same as u in the inplace case. Else we run a single evaluation of manifold to determine the size.

```julia
function g!(resid, y, p)
    resid[1:6] .= something()
end

y0 = Float64[1:11...]
cb = ManifoldProjection(g!, residual_prototype = zeros(6));
prob = ODEProblem(f!!, y0, tspan, p)
solve(prob,, callback=cb)
```

so it's just a least squares problem on the 6, that should be a better way to write this.

**Comment by abavoil** [comment]
**TL;DR**
if the initial timestep is too small, `cache.u_cache` as a random value from `similar(ũ)`, which is directly assigned to `integrator.u`. A fix is to initialize `cache.u_cache` with `ũ` instead of `similar(ũ)`.

**Longer story**
Here is a(nother) script that exhibits the bug : 

```{julia}
using OrdinaryDiffEq, DiffEqCallbacks, ADTypes, Logging

function f(du, u, p, t)
    du[1] = u[2]
    du[2] = -u[1]
end

function g(resid, u, p, t)
    resid[1] = u[2]^2 + u[1]^2 - 2
end

u0 = ones(2)
cb = ManifoldProjection(g; autodiff = AutoForwardDiff(), resid_prototype = zeros(1))
prob = ODEProblem(f, u0, 1)
prob_cb = remake(prob, callback = cb)
alg = Vern7()

with_logger(NullLogger()) do
    for tol in (1e-3, 1e-6, 1e-9, 1e-12)
        final_state_set = Set{typeof(u0)}()
        final_state_set_cb = Set{typeof(u0)}()
        for i in 1:100
            sol = solve(prob, alg, save_everystep = false, reltol = tol, abstol = tol)
            sol_cb = solve(prob_cb, alg, save_everystep = false, reltol = tol, abstol = tol)
            push!(final_state_set, sol.u[end])
            push!(final_state_set_cb, sol_cb.u[end])
        end
        println("tol: ", tol, ", no cb: ", length(final_state_set), ", cb: ", length(final_state_set_cb))
    end
end
```

It outputs this
```
tol: 0.001, no cb: 1, cb: 1
tol: 1.0e-6, no cb: 1, cb: 80
tol: 1.0e-9, no cb: 1, cb: 52
tol: 1.0e-12, no cb: 1, cb: 41
```

Low tolerances break something. Using the debugger, I found the issue. If [this statement](https://github.com/SciML/DiffEqCallbacks.jl/blob/afc84c859c258d91566859aaa54f393d6069d0d9/src/manifold.jl#L351) is reached, `cache.u_cache` may still be `similar(ũ)` from [this line](https://github.com/SciML/DiffEqCallbacks.jl/blob/afc84c859c258d91566859aaa54f393d6069d0d9/src/manifold.jl#L331). I tested 3 fixes : 
 1. Add `cache.u_cache = ũ` right before `break`ing https://github.com/SciML/DiffEqCallbacks.jl/blob/afc84c859c258d91566859aaa54f393d6069d0d9/src/manifold.jl#L349-L352,
 2. Replace `cache.ũ = u` with `cache.ũ = cache.u_cache = u` at https://github.com/SciML/DiffEqCallbacks.jl/blob/afc84c859c258d91566859aaa54f393d6069d0d9/src/manifold.jl#L300-L302
 3. Initialize `u_cache` with `ũ` instead of `similar(ũ)` at https://github.com/SciML/DiffEqCallbacks.jl/blob/afc84c859c258d91566859aaa54f393d6069d0d9/src/manifold.jl#L330-L332

The third options passes all tests.

I am figuring out how to make a PR with the fix + the relevant test : 

```{julia}
int = init(prob, dt=1e-15)
step!(int)

@assert norm(int_ok.u) ≈ 2
```
